### PR TITLE
[patch 1/4] Network rework: Rewrite client/server packet handling & packet reading

### PR DESCRIPTION
**Pull request by nerzhul**
Hi,
packet handling isn't very good, it works but it need many test at each packet receive and it's done in only one function.

This pull request create a opcode table and a handler to handle each packet properly by using a opcodehandler mapping. This improve the network code performance and read.

### CODE REVIEW

**Comment by ShadowNinja** [comment]
Seems like an improvement.  I don't know the network code well though (@sapier).

**Comment by nerzhul** [comment]
Hi @ShadowNinja i fixed tab with spaces in https://github.com/nerzhul/minetest/commit/26edd4ecfb01762d5797ab72976e5ba3d70ca882

I will work on server side handling today, and if i get time i'll look at socket handling.

**Comment by nerzhul** [comment]
@ShadowNinja in fact network code isn't handled well, the approach i propose is used since 6 year by mangos, a popular MMORPG emulator and is the best approach for a protocol handling.

I also think we must re-use unhandled opcodes instead of increasing them for future protocole usage before adding new.

This day i will work on server side. I also moved clientserver.h (which doesn't mean nothing) to networkprotocol.h, it's a better name.

Also, to have a better project view, i moved modified network stack sources to src/network/,

**Comment by nerzhul** [comment]
Ok, i rewrite server side in a proper way. Server CPU load will decrease and some useless memory traces have been cleaned.

Now i have two thing to do:

- Rewrite packet handling by using an object which have the decoding methods which contains some useful informations (datasize & data at first)
- Add some helper for some duplicated code (like player getter into packet handlers)
- Add handler for depreciated functions (need packet handling rewrite)

**Comment by nerzhul** [comment]
Hi,
i added new class, NetworkPacket, which contains the opcode, the data and some extracting methods (i'll add better methods with operators later).

Next step is to add helpers where it could be in handlers, and add those operators, and new handler for depreciated packets

**Comment by sapier** [comment]
It's quite hard to review this code while it's split. please complete your work prior doing a pull request next time!

**Comment by nerzhul** [comment]
Multiples commits permit to handle the changes, it's an agile method, then you can review the code whereas i'm writing it to improve it. It's a very very big update which permit to improve minetest network handling, and i really want it to work properly on networks :), and maybe become a minetest developper in the future, because i really love this game and i would help him to be powerful

My change isn't incompatible, because i tested a server with a new client packet handler and a client with new server packet handler, and new + new too, be careful ! no incompatible change here.

**Comment by sapier** [comment]
I already removed the label ;-) Yet well still have to verify this as there's quite some history hidden within minetest code ... sadly. 
I'm positive about this change imho it's correct direction.
Please follow coding style use spaces around operators!

**Comment by nerzhul** [comment]
Ok for coding style, i was disappointed because this code was very different in some things.

I'm working on an operator to extract and push datas to NetworkPacket, i think it's better to read packet with this method, used also by mangos.

https://github.com/mangoszero/server/blob/master/src/shared/ByteBuffer.h

**Comment by ShadowNinja** [comment]
This is coming along nicely.

In case you don't know, the code style is written down here: http://dev.minetest.net/Code_style_guidelines
A lot of the current code doesn't follow it, but we're slowly converting things.

**Comment by nerzhul** [comment]
Thanks @ShadowNinja 
for offset increment it's on the rails, i'm writing it :)

For peer_id i think it's a good idea, i don't know its use because i haven't looked where it's used, but it's a good idea. I'll do it after my rework on streaming interface

**Comment by nerzhul** [comment]
Ok @ShadowNinja i added some of your suggestions, but i am tired. I do peer_id tomorrow and i'll try to remove many stringstream to use the >> operator i choose and add some missing helpers.

I also fixed NetworkPacket destructor

**Comment by nerzhul** [comment]
Thanks for your tip @ShadowNinja i added some fixes.

**Comment by rubenwardy** [comment]
Isn't this obsolete due to the ENet patch in progress?

**Comment by nerzhul** [comment]
@rubenwardy link please ?
I haven't found a similar work

**Comment by rubenwardy** [comment]
According to discussion IRC, @sapier is rebasing this to the latest version: https://github.com/sapier/minetest/tree/enet_support

Nevermind, anyway. It would be a 0.5 change, probably. This PR is still good.

**Comment by nerzhul** [comment]
@rubenwardy right, ENet is the L4 approach, i'm on L7, that's good.

I notice some performance issues in minetest, i don't know if it's related to network but sometimes my server became at 100% CPU without anything.

I tested client loading on my 0.4.11 -stable server and client loading seems to be faster.

**Comment by rubenwardy** [comment]
The networking does need a speed boost, both in bandwidth and latency. Freeminer uses ENet, they have a much larger data bandwidth because of it. They can transfer large sound files and textures quickly - faster loading.

How much more performance does this pull request give?
Does it break compatibility with older servers?

**Comment by nerzhul** [comment]
@rubenwardy increase networking performance permit to have a better latency. For bandwidth, it's not my objective. Here it's a rewrite for improving maintenance and CPU cycle performance.

As i said in previous messages, if you look at my code:
- stable client can connect to this modified server
- modified client can connect to stable server
- modified client can connect to modified server.

Protocol hasn't change, but code was improved. Performance was gained into packet handling in one thing.
Before: look at opcode handling, it's a condition. Server and client do a IF/ELSE to match which packet it is and handle it into the ProcessData function
After: Packet opcode reference a function to handle it. It's faster, ~25-30 conditions are removed at each packet received.

And i also added a new class to handle packet like an object, not a buffer, which permit to read it easily (and in the future write it easily)

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -472,6 +475,8 @@ set(minetest_SRCS
 	sky.cpp
 	tile.cpp
 	wieldmesh.cpp
+	network/clientopcodes.cpp
+	network/toclientpacket.cpp
```
Comment (sfan5) [inline src/CMakeLists.txt]: Wouldn't it make more sense to move these up to the top to the other files you added?

File: src/client.cpp [hunk 2]
```diff
@@ -1001,1008 +1001,1091 @@ void Client::Receive()
...
-		return;
+		// Get map seed
+		*pkt >> m_map_seed;
+		infostream<<"Client: received map seed: "<<m_map_seed<<std::endl;
 	}
 
-	u8 ser_version = m_server_ser_ver;
+	if(pkt->getSize() >= 1+6+8+4)
+	{
+		// Get map seed
```
Comment (sfan5) [inline src/client.cpp]: This comment is wrong.

File: src/server.cpp [hunk 3]
```diff
@@ -1276,32 +1277,30 @@ PlayerSAO* Server::StageTwoClientInit(u16 peer_id)
 	return playersao;
 }
 
-void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
+void Server::handleCommand_Depreciated(ToServerPacket* pkt)
```
Comment (sfan5) [inline src/server.cpp]: should be `Deprecated`

File: src/server.h [hunk 4]
```diff
@@ -187,6 +188,35 @@ class Server : public con::PeerHandler, public MapEventReceiver,
 	void AsyncRunStep(bool initial_step=false);
 	void Receive();
 	PlayerSAO* StageTwoClientInit(u16 peer_id);
+	
+	/*
+	 * Command Handler
```
Comment (sfan5) [inline src/server.h]: Command Handler**s**

File: src/server.cpp [hunk 5]
```diff
@@ -1435,1301 +1437,1506 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
...
-							playeritem.getToolCapabilities(m_itemdef);
-					// Get diggability and expected digging time
-					DigParams params = getDigParams(m_nodedef->get(n).groups,
-							&playeritem_toolcap);
-					// If can't dig, try hand
-					if(!params.diggable){
-						const ItemDefinition &hand = m_itemdef->get("");
-						const ToolCapabilities *tp = hand.tool_capabilities;
-						if(tp)
-							params = getDigParams(m_nodedef->ge
```
Comment (sfan5) [inline src/server.cpp]: This is obvious, you don't need a comment for this.

File: src/server.h (hunk 4 above)
Comment (nerzhul) [inline src/server.h]: Fixed in mainline (client.h + server.h), i'll wait more comments before commit

File: src/server.cpp (hunk 5 above)
Comment (nerzhul) [inline src/server.cpp]: It's not my comment, but original comment :p, i remove it.

File: src/client.cpp (hunk 2 above)
Comment (nerzhul) [inline src/client.cpp]: Fixed mainline

File: src/server.cpp (hunk 3 above)
Comment (nerzhul) [inline src/server.cpp]: Fixed

**Comment by Zeno-** [comment]
Nice

**Comment by nerzhul** [comment]
rebase is on the way

**Comment by nerzhul** [comment]
rebase done. I'm working on memcpy improvement when creating NetworkPacket

**Comment by Zeno-** [comment]
Once this is ready, I vote for it to be merged

**Comment by nerzhul** [comment]
It's ready now. Future additions will be added in a future PR

**Comment by sapier** [comment]
please don't mixup l7 and l4 changes.
nerzhul can you fix code style in this commit? There's a lot of code left where operators aren't surrounded by spaces.

**Comment by nerzhul** [comment]
@sapier work on L4 and give me a buffer to read and set into NetworkPacket, then i handle it.
I think we must work together, maybe this week-end @sapier .

@sapier can you give me an example ?

File: src/client.cpp [hunk 6]
```diff
@@ -1001,1008 +1001,1097 @@ void Client::Receive()
...
+				<<"unsupported ser_fmt_ver"<<std::endl;
 		return;
 	}
 
-	u8 ser_version = m_server_ser_ver;
+	m_server_ser_ver = deployed;
 
-	if(command == TOCLIENT_INIT)
+	// Get player position
+	v3s16 playerpos_s16(0, BS*2+BS*20, 0);
```
Comment (sapier) [inline src/client.cpp]: missing spaces

File: src/client.cpp [hunk 7]
```diff
@@ -1001,1008 +1001,1097 @@ void Client::Receive()
...
+	if(pkt->getSize() >= 1+6)
```
Comment (sapier) [inline src/client.cpp]: missing spaces

File: src/client.cpp [hunk 8]
```diff
@@ -1001,1008 +1001,1097 @@ void Client::Receive()
...
 	{
-		if(datasize < 3)
-			return;
+		*pkt >> playerpos_s16;
+	}
+	v3f playerpos_f = intToFloat(playerpos_s16, BS) - v3f(0, BS/2, 0);
```
Comment (sapier) [inline src/client.cpp]: missing spaces

**Comment by nerzhul** [comment]
@sapier i rework on old code coding style. It was fixed

File: src/client.cpp [hunk 9]
```diff
@@ -1001,1008 +1001,1077 @@ void Client::Receive()
...
-		MapBlock *block;
+	*pkt >> time_of_day;
 
-		v2s16 p2d(p.X, p.Z);
-		sector = m_env.getMap().emergeSector(p2d);
+	time_of_day      = time_of_day % 24000;
+	float time_speed = 0;
 
-		assert(sector->getPos() == p2d);
+	if(pkt->getSize() >= 2+4) {
```
Comment (sapier) [inline src/client.cpp]: space missing

File: src/client.cpp [hunk 10]
```diff
@@ -1001,1008 +1001,1077 @@ void Client::Receive()
...
+		m_last_time_of_day_f         = time_of_day_f;
+		float time_diff            = m_time_of_day_update_timer;
+		m_time_of_day_update_timer = 0;
 
-		/*
-			Add it to mesh update queue and set it to be acknowledged after update.
-		*/
-		addUpdateMeshTaskWithEdge(p, true);
+		if(m_time_of_day_set){
+			time_speed = (3600.0*24.0) * tod_diff_f / time_diff;
```
Comment (sapier) [inline src/client.cpp]: space missing

File: src/server.cpp [hunk 11]
```diff
@@ -1309,125 +1308,128 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
...
+	/*
+		Read and check network protocol version
+	*/
 
-		/*
-			Read and check network protocol version
-		*/
+	u16 min_net_proto_version = 0;
+	if(pkt->getSize() >= 1 + PLAYERNAME_SIZE + PASSWORD_SIZE + 2)
+		min_net_proto_version = pkt->getU16(1+PLAYERNAME_SIZE+PASSWORD_SIZE);
```
Comment (sapier) [inline src/server.cpp]: space missing

File: src/server.cpp [hunk 12]
```diff
@@ -1435,1308 +1437,1487 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
...
+	// Get password
+	char given_password[PASSWORD_SIZE];
+	if(pkt->getSize() < 1 + PLAYERNAME_SIZE + PASSWORD_SIZE)
+	{
+		// old version - assume blank password
+		given_password[0] = 0;
+	}
+	else
+	{
+		for(u32 i=0; i<PASSWORD_SIZE-1; i++)
```
Comment (sapier) [inline src/server.cpp]: space missing

File: src/server.cpp [hunk 13]
```diff
@@ -1435,1308 +1437,1487 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
...
-			if(m_script->on_prejoinplayer(playername, addr_s, reason))
-			{
-				actionstream<<"Server: Player with the name \""<<playername<<"\" "
-						<<"tried to connect from "<<addr_s<<" "
-						<<"but it was disallowed for the following reason: "
-						<<reason<<std::endl;
-				DenyAccess(peer_id, narrow_to_wide(reason));
-				return;
-			}
+			given_password[i] = pkt->getChar(21+i);
```
Comment (sapier) [inline src/server.cpp]: space missing

File: src/server.cpp [hunk 14]
```diff
@@ -1435,1308 +1437,1487 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
...
 		}
+		given_password[PASSWORD_SIZE-1] = 0;
```
Comment (sapier) [inline src/server.cpp]: space missing

File: src/server.cpp [hunk 15]
```diff
@@ -1435,1308 +1437,1487 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
...
-					L"a minute.");
-		}
+	m_clients.setPlayerName(pkt->getPeerId(), playername);
 
-		m_clients.setPlayerName(peer_id,playername);
+	/*
+		Answer with a TOCLIENT_INIT
+	*/
+	{
+		SharedBuffer<u8> reply(2+1+6+8+4);
```
Comment (sapier) [inline src/server.cpp]: space missing

File: src/server.cpp [hunk 16]
```diff
@@ -1435,1308 +1437,1487 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
...
+		writeU16(&reply[0], TOCLIENT_INIT);
+		writeU8(&reply[2], deployed);
+		//send dummy pos for legacy reasons only
+		writeV3S16(&reply[2+1], floatToInt(v3f(0,0,0), BS));
```
Comment (sapier) [inline src/server.cpp]: space missing following lines too

File: src/server.cpp [hunk 17]
```diff
@@ -1435,1308 +1437,1487 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
...
+		[3+6] v3s16 pos_1
+		...
+	*/
+
+	u8 count;
+	*pkt >> count;
+
+	RemoteClient *client = getClient(pkt->getPeerId());
+
+	for(u16 i=0; i<count; i++)
```
Comment (sapier) [inline src/server.cpp]: space missing

File: src/server.cpp [hunk 18]
```diff
@@ -1435,1308 +1437,1487 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
...
+	{
+		if((s16)pkt->getSize() < 1+(i+1)*6)
```
Comment (sapier) [inline src/server.cpp]: space missing

File: src/server.cpp [hunk 19]
```diff
@@ -1435,1308 +1437,1487 @@ void Server::ProcessData(u8 *data, u32 datasize, u16 peer_id)
...
+		*pkt >> p;
+
+		client->GotBlock(p);
 	}
-	else if(command == TOSERVER_RECEIVED_MEDIA) {
+}
+
+void Server::handleCommand_PlayerPos(ToServerPacket* pkt)
+{
+	if(pkt->getSize() < 12+12+4+4)
```
Comment (sapier) [inline src/server.cpp]: space missing ... stopping to check here please check following code yourself

**Comment by sapier** [comment]
seems to be quite fine. As this is a change in critical code area I requested VanessaE to run it on her servers. If she doesn't complain about errors till January 23rd I'd suggest merging it.

**Comment by nerzhul** [comment]
@sapier ok thanks, i stay on this topic to listen to any change to do.
I'm working on packet writing and send cleanup in a second PR.

**Comment by Zeno-** [comment]
@sapier This patch is also applied and running on mtzeno.com (30000)

**Comment by nerzhul** [comment]
Rebased to current master

**Comment by Zeno-** [comment]
This has been running on my server for almost 2 weeks and I haven't noticed any issues and haven't received any reports. The code looks fine.

+1 for merging as soon as feature freeze is over

**Comment by nerzhul** [comment]
Hi Zeno, it was rebased to prevent conflict with my recent security fix

**Comment by Zeno-** [comment]
Rebase is good ;)

**Comment by Zeno-** [comment]
commit a704c04f00bfea4b77550169fa08105c2ee0dfd0

**Pull request closed without merging**